### PR TITLE
Streamline SslStream state validation

### PR DESCRIPTION
**Pull request by benaadams**
`CheckThrow` and `ThrowIfExceptional` are called on almost every op. Streamline the fast-path to and remove branches to mitigate the cost.

Pre

image

Post

image

/cc @stephentoub

### CODE REVIEW

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 1]
```diff
@@ -766,55 +770,117 @@ public override void Write(byte[] buffer, int offset, int count)
...
+                    throw new ObjectDisposedException(nameof(SslStream));
+                }
+
+                // Throw the stored exception.
+                e.Throw();
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void CheckThrowAndAuthState()
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Nit: ThrowIfExceptionalOrNotAuthenticated?

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs [hunk 2]
```diff
@@ -766,55 +770,117 @@ public override void Write(byte[] buffer, int offset, int count)
...
+            ThrowIfExceptional();
+
+            if (!IsAuthenticated)
+            {
+                ThrowNotAuthenticated();
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void CheckThrowAuthAndShutdownState()
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Nit: ThrowIfExceptionalOrNotAuthenticatedOrShutdown?

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 1 above)
Comment (benaadams) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Done

File: src/System.Net.Security/src/System/Net/Security/SslStream.cs (hunk 2 above)
Comment (benaadams) [reply src/System.Net.Security/src/System/Net/Security/SslStream.cs]: Done

**Pull request merged**